### PR TITLE
Fiqare perseo-fe improvements

### PR DESCRIPTION
**Pull request by fiqare-emergya-dev**
This pull request contains the improvements made by the Emergya team for perseo-fe. These improvements are part of the fiQare project, which is based on ISO 25010 to improve software quality. More info: https://fiqare.eu/

Changes related to variable changes have been removed in this PR.

Swagger is provided for HTTP protocol in:
<server_host>:9090/api-docs

    Note: npm install is needed

### CODE REVIEW

File: .travis.yml [hunk 1]
```diff
@@ -8,7 +8,8 @@ node_js:
 
 branches:
   only:
-    - master
+    #- master
+    - fiqare-improvements-1
```
Comment (fgalan) [inline .travis.yml]: This should be reverted (so .travis.yml ends with no change in this PR at the end)

File: .snyk [hunk 2]
```diff
@@ -0,0 +1,4 @@
+# Snyk (https://snyk.io) policy file, patches or ignores known vulnerabilities.
```
Comment (fgalan) [inline .snyk]: What is this file?

File: config.js [hunk 3]
```diff
@@ -29,7 +29,7 @@ var config = {};
  * Default log level. Can be one of: 'DEBUG', 'INFO', 'WARN', 'ERROR', 'FATAL'
  * @type {string}
  */
-config.logLevel = 'INFO';
+config.logLevel = 'DEBUG';
```
Comment (fgalan) [inline config.js]: Please rollback changes in this file. Except in the case of very good reasons (is there any?) reference configuration here shouldn't be changed.

File: lib/models/entitiesStore.js [hunk 4]
```diff
@@ -52,6 +52,9 @@ function findSilentEntities(service, subservice, ruleData, func, callback) {
         } catch (e) {
             return callback(e, null);
         }
+    } else {
+    	//Added default else clause
+        logger.debug('findSilentEntities() - Default else clause');
```
Comment (fgalan) [inline lib/models/entitiesStore.js]: Review indentation please (`//` is not aligend with `logger`)

File: lib/models/postAction.js [hunk 5]
```diff
@@ -58,6 +62,10 @@ function doIt(action, event, callback) {
         } else if (options.text) {
             requestOptions.body = options.text;
         }
+        else {
+        	//Added default else clause
+            logger.debug('doIt() - Default else clause');
```
Comment (fgalan) [inline lib/models/postAction.js]: Review indent.

File: lib/routes/actionsController.js [hunk 6]
```diff
@@ -113,7 +113,6 @@ function PostAction(req, resp) {
     } else {
         delay = config.slaveDelay;
     }
-    // TODO: Tricky ... To refactor?
```
Comment (fgalan) [inline lib/routes/actionsController.js]: Why this comment has been removed? Probably other developer put this here for some reason ;)

File: package.json [hunk 7]
```diff
@@ -20,14 +20,14 @@
   },
   "scripts": {
     "clean": "rm -rf package-lock.json && rm -rf node_modules && rm -rf coverage",
-    "test": "mocha --recursive 'test/**/*.js' --reporter spec --timeout 3000 --ui bdd --exit",
```
Comment (fgalan) [inline package.json]: Why this change?

File: package.json [hunk 8]
```diff
@@ -20,14 +20,14 @@
...
+    "test": "mocha --recursive --reporter spec --timeout 3000 --ui bdd --exit",
     "test:watch": "npm run test -- -w ./lib",
     "lint": "jshint lib/ --config .jshintrc && jshint test/ --config test/.jshintrc",
     "lint:md": "remark 'README.md' documentation",
     "lint:text": "textlint 'README.md' 'documentation/*.md' 'documentation/**/*.md'",
     "prettier": "prettier --config .prettierrc.json --write ./**/**/*.js **/*.js *.js",
     "prettier:text": "prettier 'README.md' 'documentation/*.md' 'documentation/**/*.md' --no-config --tab-width 4 --print-width 120 --write --prose-wrap always",
-    "test:coverage": "istanbul cover _mocha -- --recursive 'test/**/*.js' --reporter spec --exit",
```
Comment (fgalan) [inline package.json]: Why this change?

File: package.json [hunk 9]
```diff
@@ -52,7 +52,8 @@
     "textlint-rule-common-misspellings": "~1.0.1",
     "textlint-rule-no-dead-link": "~4.4.1",
     "textlint-rule-terminology": "~1.1.30",
-    "textlint-rule-write-good": "~1.6.2"
+    "textlint-rule-write-good": "~1.6.2",
+    "snyk": "^1.145.0"
```
Comment (fgalan) [inline package.json]: Similar comments to the one provided in https://github.com/telefonicaid/iotagent-ul/pull/412/files#r358410133 also apply to the packages.json file in this PR.

File: lib/models/notices.js [hunk 10]
```diff
@@ -249,7 +249,7 @@ function processCBv2Notice(service, subservice, ncr, ix) {
         var attrList = ncr.data[ix];
         for (var attr in attrList) {
             // Exclude id and type. NGSIv2
-            if (attr === 'id' || attr === 'type') {
+            if (attr !== 'id' && attr !== 'type') {
```
Comment (fgalan) [inline lib/models/notices.js]: Expression `(attr === 'id' || attr === 'type')` is not equivlante to `(attr !== 'id' && attr !== 'type')`.

Or instance if attr is equal to "id" then:

```
(attr === 'id' || attr === 'type') =>  (True || False) => True
```

while

```
(attr !== 'id' && attr !== 'type') => (False && True) => False
```

File: .snyk (hunk 2 above)
Comment (fiqare-emergya-dev) [reply .snyk]: This file is a file generated by a tool used during the analyses. I'm going to delete it.

File: .travis.yml (hunk 1 above)
Comment (fiqare-emergya-dev) [reply .travis.yml]: I'm going to revert the changes, sorry.

File: config.js (hunk 3 above)
Comment (fiqare-emergya-dev) [reply config.js]: It's a mistake, sorry.

File: lib/routes/actionsController.js (hunk 6 above)
Comment (fiqare-emergya-dev) [reply lib/routes/actionsController.js]: This comment has been deleted due to one of the rules of the maintainability axis (ISO 25010).

TODO tags are commonly used to mark places where some more code is required, but which the developer wants to implement later.
Sometimes the developer will not have the time or will simply forget to get back to that tag.
This rule is meant to track those tags and to ensure that they do not go unnoticed.

More info: http://cwe.mitre.org/data/definitions/546.html

File: package.json (hunk 9 above)
Comment (fiqare-emergya-dev) [reply package.json]: This is a tool used during the analyses. I'm going to delete it.

File: package.json (hunk 7 above)
Comment (fiqare-emergya-dev) [reply package.json]: This is a mistake in the merge, sorry.

File: package.json (hunk 8 above)
Comment (fiqare-emergya-dev) [reply package.json]: This is a mistake in the merge, sorry.

File: lib/routes/actionsController.js (hunk 6 above)
Comment (fgalan) [reply lib/routes/actionsController.js]: In the case of not using a TODO marck, which other mechanism do you propose to track the need of a refactor mentioned by the developer here?

Comment (fiqare-emergya-dev) [reply lib/routes/actionsController.js]: I have reviewed it and it is an issue of INFO level, it is just a suggestion. Sorry for the inconvenience.

Comment (fgalan) [reply lib/routes/actionsController.js]: Don't worry ;)

Thus, I understand the change will be rolled-back.

File: .travis.yml (hunk 1 above)
Comment (fiqare-emergya-dev) [reply .travis.yml]: Fixed in 08201e79ae90920e0f9707b01629f3f281c7d4ad

File: .snyk (hunk 2 above)
Comment (fiqare-emergya-dev) [reply .snyk]: Fixed in 08201e79ae90920e0f9707b01629f3f281c7d4ad

File: config.js (hunk 3 above)
Comment (fiqare-emergya-dev) [reply config.js]: Fixed in 08201e79ae90920e0f9707b01629f3f281c7d4ad

File: package.json (hunk 7 above)
Comment (fiqare-emergya-dev) [reply package.json]: Fixed in 08201e79ae90920e0f9707b01629f3f281c7d4ad

File: package.json (hunk 8 above)
Comment (fiqare-emergya-dev) [reply package.json]: Fixed in 08201e79ae90920e0f9707b01629f3f281c7d4ad

File: package.json (hunk 9 above)
Comment (fiqare-emergya-dev) [reply package.json]: Fixed in 08201e79ae90920e0f9707b01629f3f281c7d4ad

File: lib/models/notices.js (hunk 10 above)
Comment (fiqare-emergya-dev) [reply lib/models/notices.js]: Fixed in 08201e79ae90920e0f9707b01629f3f281c7d4ad

File: lib/routes/actionsController.js (hunk 6 above)
Comment (fiqare-emergya-dev) [reply lib/routes/actionsController.js]: Fixed in 3020aab9fb0f1a72ce0254cc6bc7db8401d13359

File: config.js (hunk 3 above)
Comment (fgalan) [reply config.js]: Are you sure? Looking to the diff tab I'm still seeing changes in config.js file ...

File: package.json [hunk 11]
```diff
@@ -65,10 +65,12 @@
     "nodemailer": "~1.11.0",
     "nodemailer-smtp-transport": "~0.1.13",
     "request": "2.88.0",
+    "smpp": "0.3.1",
+    "swagger-jsdoc": "^3.4.0",
```
Comment (fgalan) [inline package.json]: ^ should be changed to ~ in these dependencies

File: lib/routes/visualRulesController.js [hunk 12]
```diff
@@ -127,7 +127,6 @@ function PutVR(req, resp) {
             return myutils.respond(resp, err, data);
         }
         if (data && data.name) {
-            //resp.location(req.url.replace(new RegExp(req.params.id), '')+encodeURIComponent(data.name));
```
Comment (fgalan) [inline lib/routes/visualRulesController.js]: @AlvaroVega maybe this is an old leftover? It's a good idea to remove it?

File: documentation/eslint.md [hunk 13]
```diff
@@ -0,0 +1,74 @@
+# Development documentation (Third Iteration)
```
Comment (fgalan) [inline documentation/eslint.md]: Not sure of understanding the purpose of this file...

Apart from the title (what is "third iteration"? maybe it makes sense in the context of fiQare project, but not this repo is project-agnostic), this seem to be just a description of eslint based on standard content.

I mean, I could understand a replacmente of jslint with eslint, as for instance is done by PR https://github.com/telefonicaid/iotagent-ul/pull/408 in the case of IOTA-UL (although not in the scope of this PR but in a new one). However just including a piece of documentation about eslint in the repository doesn't have too much sense to me.

Maybe I'm not getting the point of something. Could you elaborate on this? If it is easier for you, you can tell me by email.

File: package.json (hunk 11 above)
Comment (fiqare-emergya-dev) [reply package.json]: Fixed in 51271cd82fd35f866a09fed1f4debe74380050a0

File: documentation/eslint.md (hunk 13 above)
Comment (vladiricoemergya) [reply documentation/eslint.md]: This is a style guide made for the Fiqare project, if it is not useful for Perseo Front or for future developers we remove it from the PR.

Comment (fiqare-emergya-dev) [reply documentation/eslint.md]: This is a style guide made for the Fiqare project, if it is not useful for Perseo Front or for future developers we remove it from the PR.

Comment (fgalan) [reply documentation/eslint.md]: Yes, it seems to be too much fiQare-specific... Removing it from the PR sounds good to me.

Comment (fiqare-emergya-dev) [reply documentation/eslint.md]: Fixed in ce583be19f1561114b48d1dc71d91e661e2e1b11

**Comment by fgalan** [comment]
At the present moment (January 7th, 2019), there are some comment threads (at least three) not yet answered/fixed so I understand the work on this PR is ongoing. 

@fiqare-emergya-dev, is my understanding correct?

File: config.js (hunk 3 above)
Comment (fiqare-emergya-dev) [reply config.js]: I've reviewed the commit and it's corrected.
https://github.com/fiqare-emergya-dev/cdti-fiqare-perseo-fe-src/blob/fiqare-improvements-1/config.js

File: lib/models/entitiesStore.js (hunk 4 above)
Comment (fiqare-emergya-dev) [reply lib/models/entitiesStore.js]: I checked the last commit and I can't locate the indentation error
https://github.com/fiqare-emergya-dev/cdti-fiqare-perseo-fe-src/blob/fiqare-improvements-1/lib/models/entitiesStore.js

File: lib/models/postAction.js (hunk 5 above)
Comment (fiqare-emergya-dev) [reply lib/models/postAction.js]: Fixed in d978818056a0293666063e39270979bdb5adb8ab

**Comment by fiqare-emergya-dev** [comment]
I've already reviewed the comment threads.

File: config.js (hunk 3 above)
Comment (fgalan) [reply config.js]: With regards to this particular change, that's right, it has been rolled-back.

However, the PR stills containing changes in config.js. I see the following in the "Diff" tab:

File: lib/models/entitiesStore.js (hunk 4 above)
Comment (fgalan) [reply lib/models/entitiesStore.js]: The diff shows the miss-indentation:

imagen

Maybe it's someking of tabs vs. space issue. Tabs shouldn't be used in Perseo FE code, according to our code style.

Comment (vladiricoemergya) [reply lib/models/entitiesStore.js]: Fixed in ff07d1070d5096cdbc0e359130c6a04514f87ea7

File: config.js (hunk 3 above)
Comment (fiqare-emergya-dev) [reply config.js]: Fixed in ff07d1070d5096cdbc0e359130c6a04514f87ea7

File: lib/models/entitiesStore.js (hunk 4 above)
Comment (fiqare-emergya-dev) [reply lib/models/entitiesStore.js]: Fixed in 0865dc81b89788cff47bcc99fb97d808534312bb

File: lib/models/emailAction.js [hunk 14]
```diff
@@ -25,13 +25,12 @@
...
     nodemailer = require('nodemailer'),
     logger = require('logops'),
-    smtpTransport = require('nodemailer-smtp-transport'),
     config = require('../../config'),
     myutils = require('../myutils'),
     alarm = require('../alarm'),
     metrics = require('./metrics');
 
-var transporter = nodemailer.createTransport(smtpTransport(config.smtp));
+var transporter = nodemailer.createTransport(config.smtp);
```
Comment (fgalan) [inline lib/models/emailAction.js]: Why this change (along with the removal in L28) is actually needed?

File: lib/models/emailAction.js [hunk 15]
```diff
@@ -50,7 +49,7 @@ function SendMail(action, event, callback) {
 
         metrics.IncMetrics(event.service, event.subservice, metrics.actionEmail);
 
-        transporter.sendMail(mailOptions, function(err, info) {
+        transporter.sendMail(mailOptions, (err, info) => {
```
Comment (fgalan) [inline lib/models/emailAction.js]: I understand the preferred code style is currently `function(error,info)` over `(err, info) =>`, so this change probably should be rolled-back.

File: lib/models/rules.js [hunk 16]
```diff
@@ -33,7 +33,8 @@ var util = require('util'),
     actions = require('./actions'),
     namePattern = /^[a-zA-Z0-9_-]+$/,
     MaxNameLength = 50,
-    errors = {};
+    errors = {},
+    logger = require('logops');
```
Comment (fgalan) [inline lib/models/rules.js]: Better to have all the require() together. Pleace move up (just afeter `actions` declaration).

File: lib/perseo.js [hunk 17]
```diff
@@ -98,6 +100,22 @@ function start(callbackStart) {
         versionRoutes.AddTo(app);
         metricsRoutes.AddTo(app);
 
+        const options = {
+            definition: {
+                swagger: '2.0', // Specification (optional, defaults to swagger: '2.0')
+                info: {
+                    title: 'Perseo Front-End', // Title (required)
+                    version: '1.7.0-fiqare' // Version (required)
```
Comment (fgalan) [inline lib/perseo.js]: This should be got from packages.json

File: package.json [hunk 18]
```diff
@@ -65,10 +65,12 @@
     "nodemailer": "~1.11.0",
     "nodemailer-smtp-transport": "~0.1.13",
     "request": "2.88.0",
+    "smpp": "0.3.1",
+    "swagger-jsdoc": "~3.4.0",
```
Comment (fgalan) [inline package.json]: documentation/development.md documentation should be modified to include a section about swagger (similar to what it has been done i n https://github.com/telefonicaid/iotagent-ul/pull/412)

**Comment by fgalan** [comment]
I have done a new review round. Please, find my new comments in the PR.

In addition to them, a couple of additional comments:

* It would be great if you could squash & rebase your PR to pass from 107 commits to 1 commit (or a small number)
* Review Travis CI results, as it seems the PR is not passing the tests

Thanks!

File: lib/models/emailAction.js (hunk 14 above)
Comment (fiqare-emergya-dev) [reply lib/models/emailAction.js]: The function createTransport accepts the configuration as a string, so the deleted var in L28 was no necessary and redundant.

Comment (fgalan) [reply lib/models/emailAction.js]: Ok. NTC.

File: lib/models/emailAction.js (hunk 15 above)
Comment (fiqare-emergya-dev) [reply lib/models/emailAction.js]: Fixed in 9c80eb9

File: lib/models/rules.js (hunk 16 above)
Comment (fiqare-emergya-dev) [reply lib/models/rules.js]: Fixed in 9c80eb9

File: lib/perseo.js (hunk 17 above)
Comment (fiqare-emergya-dev) [reply lib/perseo.js]: Fixed in 9c80eb9

File: package.json (hunk 18 above)
Comment (fiqare-emergya-dev) [reply package.json]: readme.md updated in 9c80eb9

File: README.md [hunk 19]
```diff
@@ -49,3 +49,28 @@ wish to make a clarifying public statement as follows:
 > incorporate enhancements is considered a derivative work of the product. Software that merely uses or aggregates (i.e.
 > links to) an otherwise unmodified version of existing software is not considered a derivative work, and therefore it
 > does not need to be released as under the same license, or even released as open source.
+
+
+## User & Programmers Manual
```
Comment (fgalan) [inline README.md]: I think README.md file is not the right place to include this. In addition, it should be simplified: you have already included the needed packages in packages.json and so on, so please focus just in the usage.

In sum, It should be included in development.md, just after the last item (Prettify Code) and be like this:

```
### Swagger

In order to run Swagger, you need to execute the Perseo FE (as explained [here](deployment.md) and then you
can access to:

```
<server_host>:9090/api-docs
```

The swagger documentation provided at /api-docs covers all the HTTP endpoint exposed by Perseo FE.
```

File: lib/models/rulesStore.js [hunk 20]
```diff
@@ -97,7 +97,7 @@ function search(rule, callback) {
 
 function findAll(service, subservice, callback) {
     var criterion = {};
-
+
```
Comment (fgalan) [inline lib/models/rulesStore.js]: Remove this extra whitespace, pls.

File: lib/perseo.js [hunk 21]
```diff
@@ -48,7 +48,10 @@ var domain = require('domain'),
     serviceMiddleware = require('./middleware/service'),
     errorMiddleware = require('./middleware/error'),
     myutils = require('./myutils'),
-    d = domain.create();
+    d = domain.create(),
+    swaggerJsdoc = require('swagger-jsdoc'),
+    swaggerUi = require('swagger-ui-express'),
+    pjson = require('./package.json');
```
Comment (fgalan) [inline lib/perseo.js]:
```suggestion
    pjson = require('../package.json');
```

Not sure, please check.

File: lib/routes/visualRulesController.js (hunk 12 above)
Comment (AlvaroVega) [reply lib/routes/visualRulesController.js]: I guess it could be deleted

File: lib/alarm.js [hunk 22]
```diff
@@ -30,6 +30,7 @@ var util = require('util'),
 function raise(alarm, context, message) {
     var state = alarms[alarm];
     context = (process.domain && process.domain.context) || {};
+
```
Comment (fgalan) [inline lib/alarm.js]: CHANGES_NEXT_RELEASE entries should be added regarding the changes in this documentation. Suggestion:

```
 - Add: /api-docs endpoint providing swagger-based documentation of the HTTP endpoints exposed by Perseo FE
- Hardening: software quality improvement based on ISO25010 recommendations 
```

File: lib/routes/visualRulesController.js (hunk 12 above)
Comment (fgalan) [reply lib/routes/visualRulesController.js]: Thus, change is fine. NTC.

File: lib/alarm.js (hunk 22 above)
Comment (fiqare-emergya-dev) [reply lib/alarm.js]: Fixed in 210813ba445911619199c59d729c8b724f5aef5a

File: lib/models/rulesStore.js (hunk 20 above)
Comment (fiqare-emergya-dev) [reply lib/models/rulesStore.js]: Fixed in 210813ba445911619199c59d729c8b724f5aef5a

File: README.md (hunk 19 above)
Comment (fiqare-emergya-dev) [reply README.md]: Fixed in 210813ba445911619199c59d729c8b724f5aef5a

File: lib/perseo.js (hunk 21 above)
Comment (fiqare-emergya-dev) [reply lib/perseo.js]: it is correct

Comment (fgalan) [reply lib/perseo.js]: Ok. Fixed in df81a41f6f55ee0023a29090136c24d0955ca20a

File: test/component/metrics/metrics_actions_test.js [hunk 23]
```diff
@@ -226,13 +226,13 @@ describe('Metrics', function() {
 
                                 should.equal(m.services.unknownt.sum.actionEntityUpdate, 1);
                                 should.equal(m.services.unknownt.sum.okActionEntityUpdate, 0);
-                                should.equal(m.services.unknownt.sum.failedActionEntityUpdate, 1);
+                                should.equal(m.services.unknownt.sum.failedActionEntityUpdate, 0);
```
Comment (AlvaroVega) [inline test/component/metrics/metrics_actions_test.js]: This is going to be handled in another PR: https://github.com/telefonicaid/perseo-fe/pull/425

File: lib/models/updateAction.js [hunk 24]
```diff
@@ -188,8 +188,9 @@ function doRequest(action, event, token, callback) {
             }
             if (!err && data.body && data.body.orionError) {
                 err = new errors.OrionError(JSON.stringify(data.body.orionError));
+            } else {
+                callback(err, data);
             }
-            callback(err, data);
```
Comment (fgalan) [inline lib/models/updateAction.js]: I think this change is no correct... the callback() needs to be changed in any case and, after this change, it is only called when the execution flow goes by the "else" part.

File: test/component/metrics/metrics_actions_test.js (hunk 23 above)
Comment (fgalan) [reply test/component/metrics/metrics_actions_test.js]: In order not disturb @fiqare-emergya-dev (a conflict will happen when #425 gets merged), let's keep as it is now in this PR. I'll fix it in the prelanding branch (hardening/fiqare-perseo-fe-improvements-prelanding)

NTC (just informative)

File: lib/models/updateAction.js (hunk 24 above)
Comment (fiqare-emergya-dev) [reply lib/models/updateAction.js]: Fixed in c29e89f6dc2c5b39008fa8a025d1cf01573c17e3
